### PR TITLE
WildCard support in origins

### PR DESCRIPTION
**Pull request by alireza-naghizadeh**
Ability to allow using wildcard in origins...
Example (like DNS & IIS.10 bindings):

```
app.UseCors(policy => policy
    policy.WithOrigins("http://example.com", "http://*.example.com")
);
```

@dnfclas

### CODE REVIEW

**Comment by alireza-naghizadeh** [comment]
The better example (like DNS & IIS.10 bindings):

app.UseCors(policy => policy
policy.WithOrigins("http://example.com", "http://*.example.com")
);

@dnfclas

**Comment by danroth27** [comment]
Confirmed the CLA was signed

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs [hunk 1]
```diff
@@ -82,7 +82,7 @@ public CorsResult EvaluatePolicy(HttpContext context, CorsPolicy policy)
         public virtual void EvaluateRequest(HttpContext context, CorsPolicy policy, CorsResult result)
         {
             var origin = context.Request.Headers[CorsConstants.Origin];
-            if (StringValues.IsNullOrEmpty(origin) || !policy.AllowAnyOrigin && !policy.Origins.Contains(origin))
+            if (StringValues.IsNullOrEmpty(origin) || !policy.AllowAnyOrigin && !policy.RegularOrigins.Any(r => r.IsMatch(origin)))
```
Comment (troydai) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs]: What's the performance implication here? The impact could be big.

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicy.cs [hunk 2]
```diff
@@ -141,5 +149,23 @@ public override string ToString()
...
             return builder.ToString();
         }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="CorsPolicy"/>.
+        /// </summary>
+        public CorsPolicy()
+        {
+            // Regenerate 'RegularOrigins' every time the 'Origins' has changed.
+            Origins.CollectionChanged += (sender, e) =>
```
Comment (troydai) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicy.cs]: Are policies all generated and set during configuration? If so using a `ObservableCollection` seems to be an overkill. Are there mechanism to update patterns once when all the origins are added?

Comment (alireza-naghizadeh) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicy.cs]: > Are policies all generated and set during configuration?

No! We can change origins out of the configuration... When you need dynamic origins, you can write a custom service divided from `CorsService`... I have a project that I change origins dynamically in `EvaluateRequest` & `EvaluatePreflightRequest` with access to `policy.Origins`...
So, by using `ObservableCollection` we can generate `RegularOrigins` each time is needed.

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicy.cs]: We shouldn't use an observable collection, instead an interface with a method call would be better than having code react to an event handlers whenever the collection changes.

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs]: Regex by default? I don't think so 😄

Comment (alireza-naghizadeh) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs]: We can improve performance by check both (combined), by default check origin strings and if needed wildcard then check regular origins...

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs]: Sure, anything with regex raises a red flag. We avoid them like the plague and performance results need to be shown whenever they are used. They also need a timeout so @blowdart doesn't freak out 😄

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicy.cs (hunk 2 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicy.cs]: I also don't see how this is related to the PR. Can you make this change only about wildcards and separate dynamic configuration in another issue/PR.

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs]: Actually why do you even need to support regex? Just do the wildcard matching by hand, it'll be more predictable and efficient that way. Also we can start by allowing less things. Opening it up to regex makes it hard to optimize.

**Comment by alireza-naghizadeh** [comment]
I changed the codes! Now there isn't any Regex...
I was solving many issues (dynamic origins, regular support...), but @davidfowl is right, Now we should solve this issue (WildCard support)...
Thanks

**Comment by Eilon** [comment]
We merged https://github.com/aspnet/CORS/pull/94, so closing this.

**Pull request closed without merging**